plugins/org.eclipse.xtext.builder.standalone/src/org/eclipse/xtext/builder/standalone/incremental/ChunkedResourceDescriptions.xtend

```diff
@@ -0,0 +1,71 @@
+/*******************************************************************************
+ * Copyright (c) 2015 itemis AG (http://www.itemis.eu) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.builder.standalone.incremental
+
+import java.util.HashMap
+import java.util.Map
+import org.eclipse.emf.common.util.URI
+import org.eclipse.xtext.resource.IResourceDescriptions
+import org.eclipse.xtext.resource.impl.AbstractCompoundSelectable
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsData
+
+/**
+ * 
+ * 
+ * @author Sven Efftinge - Initial contribution and API
+ * 
+ * @since 2.9
+ */
+class ChunkedResourceDescriptions extends AbstractCompoundSelectable implements IResourceDescriptions {
+	
+	protected volatile Map<String, ResourceDescriptionsData> chunk2resourceDescriptions = newHashMap;
+	
+	new() {}
+	
+	new(Map<String,ResourceDescriptionsData> initialData) {
+		this.chunk2resourceDescriptions = new HashMap(initialData) 
+	}
+	
+	def ResourceDescriptionsData setContainer(String name, ResourceDescriptionsData descriptions) {
+		val copy = new HashMap(chunk2resourceDescriptions) 
```

Comment:
That was also my first idea, but I guess we'll see concurrentModificationException when the map is changed while values() are traversed.